File: script/server-side/controllers/admin_controller.lua
-- =============================================================
--  mri_esc — Admin Controller
-- =============================================================

local adminCooldowns = {}

-- ─────────────────────────────────────────────────────────────
--  ADMIN: LISTAR VIPs
-- ─────────────────────────────────────────────────────────────
lib.callback.register('mri_esc:vip:admin:list', function(source)
    if not IsAdminPlayer(source) then return { list = {}, stats = { total = 0, online = 0, offline = 0 } } end

    local list = {}
    local onlineCount = 0
    local offlineCount = 0

    local ok, records = pcall(function()
        return MySQL.query.await("SELECT * FROM mri_vip_records LIMIT 500")
    end)

    if ok and records then
        for _, r in ipairs(records) do
            local player = exports.qbx_core:GetPlayerByCitizenId(r.citizenid) or exports.qbx_core:GetOfflinePlayer(r.citizenid)
            if player then
                local isOnline = not player.Offline
                if isOnline then onlineCount = onlineCount + 1 else offlineCount = offlineCount + 1 end

                list[#list + 1] = {
                    citizenid      = r.citizenid,
                    name           = player.PlayerData.charinfo.firstname .. " " .. player.PlayerData.charinfo.lastname,
                    tier           = r.tier,
                    online         = isOnline,
                    source         = isOnline and player.PlayerData.source or nil,
                    granted_at     = r.granted_at,
                    expires_at     = r.expires_at,
                    total_earned   = r.total_earned or 0,
                    paycheck_count = r.paycheck_count or 0,
                    granted_by     = r.granted_by,
                }
            end
        end
    end

    -- Fallback for metadata-only VIPs
    for _, player in pairs(exports.qbx_core:GetQBPlayers()) do
        local vip = player.PlayerData.metadata['vip']
        if vip and vip ~= 'nenhum' then
            local found = false
            for _, item in ipairs(list) do
                if item.citizenid == player.PlayerData.citizenid then
                    found = true
                    break
                end
            end
            if not found then
                onlineCount = onlineCount + 1
                list[#list + 1] = {
                    citizenid = player.PlayerData.citizenid,
                    name = player.PlayerData.charinfo.firstname .. " " .. player.PlayerData.charinfo.lastname,
                    tier = vip,
                    online = true,
                    source = player.PlayerData.source,
                    granted_at = 0,
                    expires_at = nil,
                    total_earned = 0,
                    paycheck_count = 0,
                    granted_by = 'mri_qbox_metadata'
                }
            end
        end
    end

    local allPlans = {}
    if VipPlansConfigs then
        for id, cfg in pairs(VipPlansConfigs) do
            allPlans[#allPlans + 1] = { id = id, label = cfg.label, payment = cfg.payment, inventory = cfg.inventory, benefits = cfg.benefits, rewards = cfg.rewards or {} }
        end
    end

    return {
        list = list,
        allPlans = allPlans,
        stats = {
            total = onlineCount + offlineCount,
            online = onlineCount,
            offline = offlineCount
        }
    }
end)

-- ─────────────────────────────────────────────────────────────
--  ADMIN: CONCEDER VIP
-- ─────────────────────────────────────────────────────────────
lib.callback.register('mri_esc:vip:admin:grant', function(source, data)
    local srcStr = tostring(source)
    if adminCooldowns[srcStr] and (os.time() - adminCooldowns[srcStr]) < 3 then
        return { success = false, error = "Aguarde 3 segundos entre ações" }
    end
    adminCooldowns[srcStr] = os.time()

    if not IsAdminPlayer(source) then return { success = false, error = "Sem permissão" } end
    if not data or not data.citizenId or not data.tier then
        return { success = false, error = "Dados inválidos" }
    end

    local cid = data.citizenId:upper()
    local adminName = "Admin"
    pcall(function()
        local ap = exports.qbx_core:GetPlayer(source)
        if ap then
            adminName = ap.PlayerData.charinfo.firstname .. " " .. ap.PlayerData.charinfo.lastname
        end
    end)

    local success, result = false, "Erro interno"
    if GrantVip then
        success, result = GrantVip(cid, data.tier, data.durationDays, adminName)
    else
        local ok, err = pcall(function()
            local online = exports.qbx_core:GetPlayerByCitizenId(cid)
            if online then
                online.Functions.SetMetaData('vip', data.tier)
            else
                local off = exports.qbx_core:GetOfflinePlayer(cid)
                if off then
                    off.PlayerData.metadata['vip'] = data.tier
                    exports.qbx_core:SaveOffline(off.PlayerData)
                end
            end
            if MySQL and MySQL.query then
                local now = os.time()
                local exp = data.durationDays and tonumber(data.durationDays) > 0
                    and (now + tonumber(data.durationDays) * 86400) or nil
                MySQL.query([[
                    INSERT INTO mri_vip_records (citizenid, tier, granted_at, expires_at, granted_by, updated_at)
                    VALUES (?, ?, ?, ?, ?, ?)
                    ON DUPLICATE KEY UPDATE
                        tier=VALUES(tier), granted_at=VALUES(granted_at),
                        expires_at=VALUES(expires_at), granted_by=VALUES(granted_by),
                        updated_at=VALUES(updated_at)
                ]], { cid, data.tier, now, exp, adminName, now })
            end
        end)
        success = ok
        result = err
    end

    return success and { success = true } or { success = false, error = tostring(result) }
end)

-- ─────────────────────────────────────────────────────────────
--  ADMIN: REVOGAR VIP
-- ─────────────────────────────────────────────────────────────
lib.callback.register('mri_esc:vip:admin:revoke', function(source, data)
    if not IsAdminPlayer(source) then return { success = false, error = "Sem permissão" } end
    if not data or not data.citizenId then return { success = false, error = "citizenId obrigatório" } end

    local cid = data.citizenId:upper()
    pcall(function()
        if RevokeVip then
            RevokeVip(cid, 'admin')
        else
            local online = exports.qbx_core:GetPlayerByCitizenId(cid)
            if online then
                online.Functions.SetMetaData('vip', nil)
            else
                local off = exports.qbx_core:GetOfflinePlayer(cid)
                if off then
                    off.PlayerData.metadata['vip'] = nil
                    exports.qbx_core:SaveOffline(off.PlayerData)
                end
            end
            if MySQL and MySQL.query then
                MySQL.query("DELETE FROM mri_vip_records WHERE citizenid = ?", { cid })
            end
        end
    end)

    return { success = true }
end)

-- ─────────────────────────────────────────────────────────────
--  ADMIN: RENOVAR VIP
-- ─────────────────────────────────────────────────────────────
lib.callback.register('mri_esc:vip:admin:extend', function(source, data)
    if not IsAdminPlayer(source) then return { success = false, error = "Sem permissão" } end
    if not data or not data.citizenId or not data.days then return { success = false, error = "Dados inválidos" } end

    local cid = data.citizenId:upper()
    local adminName = "Admin"
    pcall(function()
        local ap = exports.qbx_core:GetPlayer(source)
        if ap then adminName = ap.PlayerData.charinfo.firstname .. " " .. ap.PlayerData.charinfo.lastname end
    end)

    local ok = false
    if ExtendVip then
        local extOk = pcall(ExtendVip, cid, data.tier, data.days, adminName)
        ok = extOk
    elseif MySQL and MySQL.query then
        ok = true
        pcall(function()
            local now  = os.time()
            local recs = MySQL.query.await("SELECT expires_at FROM mri_vip_records WHERE citizenid = ?", { cid })
            local base = (recs and recs[1] and recs[1].expires_at and recs[1].expires_at > now)
                and recs[1].expires_at or now
            local newExp = base + (tonumber(data.days) * 86400)
            MySQL.query(
                "INSERT INTO mri_vip_records (citizenid, tier, granted_at, expires_at, granted_by, updated_at) VALUES (?, ?, 0, ?, ?, ?) ON DUPLICATE KEY UPDATE tier=VALUES(tier), expires_at=VALUES(expires_at), granted_by=VALUES(granted_by), updated_at=VALUES(updated_at)",
                { cid, data.tier or 'tier1', newExp, adminName, now }
            )
        end)
    end

    return ok and { success = true } or { success = false, error = "Falha ao renovar" }
end)

-- ─────────────────────────────────────────────────────────────
--  ADMIN: BUSCA DE JOGADORES
-- ─────────────────────────────────────────────────────────────
lib.callback.register('mri_esc:vip:admin:search', function(source, data)
    if not IsAdminPlayer(source) then return {} end
    if not data or not data.query or #data.query < 2 then return {} end

    local results = {}
    local search  = data.query:lower()

    pcall(function()
        for _, player in pairs(exports.qbx_core:GetQBPlayers()) do
            local pd   = player.PlayerData
            local name = (pd.charinfo.firstname or "") .. " " .. (pd.charinfo.lastname or "")
            if name:lower():find(search, 1, true) or pd.citizenid:lower():find(search, 1, true) then
                results[#results + 1] = {
                    citizenid = pd.citizenid,
                    name      = name,
                    online    = true,
                    vip       = pd.metadata['vip'] or 'nenhum'
                }
            end
        end
    end)

    if MySQL and MySQL.query then
        local s = "%" .. data.query .. "%"
        local ok, rows = pcall(function()
            return MySQL.query.await([[
                SELECT p.citizenid,
                    CONCAT(JSON_UNQUOTE(JSON_EXTRACT(p.charinfo,'$.firstname')),' ',
                           JSON_UNQUOTE(JSON_EXTRACT(p.charinfo,'$.lastname'))) AS name,
                    JSON_UNQUOTE(JSON_EXTRACT(p.metadata,'$.vip')) AS vip
                FROM players p
                WHERE p.citizenid LIKE ?
                   OR JSON_UNQUOTE(JSON_EXTRACT(p.charinfo,'$.firstname')) LIKE ?
                   OR JSON_UNQUOTE(JSON_EXTRACT(p.charinfo,'$.lastname'))  LIKE ?
                LIMIT 10
            ]], { s, s, s })
        end)
        if ok and rows then
            for _, row in ipairs(rows) do
                local found = false
                for _, r in ipairs(results) do
                    if r.citizenid == row.citizenid then found = true; break end
                end
                if not found then
                    results[#results + 1] = {
                        citizenid = row.citizenid,
                        name      = row.name,
                        online    = false,
                        vip       = row.vip or 'nenhum'
                    }
                end
            end
        end
    end

    return results
end)

-- NOTE: getPlans, savePlan, deletePlan, getItems callbacks are registered
-- in modules/vip-manager/server/callbacks.lua (loaded first by fxmanifest).
-- DO NOT re-register them here or rewards data will be lost.

File: script/server-side/controllers/admin_controller.js
{
    const adminCooldowns = {};
    const resourceName = GetCurrentResourceName();
    const bridge = exports[resourceName];

    console.log("[vanguard_esc] ADMIN CONTROLLER JS LOADING (Full Version)...");

    const lib = {
        callback: {
            register: (name, cb) => {
                if (exports.ox_lib && exports.ox_lib.setValidCallback) {
                    exports.ox_lib.setValidCallback(name, true);
                }
                onNet(`__ox_cb_${name}`, async (resource, key, ...args) => {
                    const src = source;
                    try {
                        const result = await cb(src, ...args);
                        TriggerClientEvent(`__ox_cb_${resource}`, src, key, result);
                    } catch (err) {
                        console.error(`[vanguard_esc] Admin Callback Error [${name}]: ${err.message}`);
                        TriggerClientEvent(`__ox_cb_${resource}`, src, key, false);
                    }
                });
            }
        }
    };

    lib.callback.register('mri_esc:vip:admin:list', async (source) => {
        if (!bridge.IsAdminPlayer(source)) {
            return { list: [], stats: { total: 0, online: 0, offline: 0 } };
        }

        let list = [];
        let onlineCount = 0;
        let offlineCount = 0;

        try {
            const records = await exports.oxmysql.query_async("SELECT * FROM mri_vip_records LIMIT 500");
            console.log(`[vanguard_esc] DEBUG: Admin list SQL fetched ${records ? records.length : 0} records`);
            
            if (records && records.length > 0) {
                for (const r of records) {
                    const cid = r.citizenid || r.citizenId;
                    if (!cid) continue;
                    
                    const player = exports.qbx_core.GetPlayerByCitizenId(cid) || exports.qbx_core.GetOfflinePlayer(cid);
                    if (player) {
                        const isOnline = !player.Offline;
                        if (isOnline) onlineCount++; else offlineCount++;

                        list.push({
                            citizenid: cid,
                            name: `${player.PlayerData.charinfo.firstname} ${player.PlayerData.charinfo.lastname}`,
                            tier: r.tier || player.PlayerData.metadata['vip'] || 'nenhum',
                            online: isOnline,
                            source: isOnline ? player.PlayerData.source : null,
                            granted_at: r.granted_at || 0,
                            expires_at: r.expires_at || null,
                            total_earned: r.total_earned || 0,
                            paycheck_count: r.paycheck_count || 0,
                            granted_by: r.granted_by || 'unknown',
                        });
                    }
                }
            }
        } catch (err) {
            console.error(`[vanguard_esc] Error in admin:list SQL: ${err.message}`);
        }

        const players = exports.qbx_core.GetQBPlayers();
        const seenCids = new Set(list.map(item => item.citizenid));

        for (const key in players) {
            const player = players[key];
            const vip = player.PlayerData.metadata['vip'];
            if (vip && vip !== 'nenhum' && !seenCids.has(player.PlayerData.citizenid)) {
                onlineCount++;
                list.push({
                    citizenid: player.PlayerData.citizenid,
                    name: `${player.PlayerData.charinfo.firstname} ${player.PlayerData.charinfo.lastname}`,
                    tier: vip,
                    online: true,
                    source: player.PlayerData.source,
                    granted_at: 0,
                    expires_at: null,
                    total_earned: 0,
                    paycheck_count: 0,
                    granted_by: 'mri_qbox_metadata'
                });
            }
        }

        const safeParse = (data) => {
            if (typeof data === 'object' && data !== null) return data;
            try { return JSON.parse(data || "[]"); } catch (e) { return []; }
        };

        let allPlans = [];
        try {
            const plans = await exports.oxmysql.query_async("SELECT * FROM mri_vip_plans");
            if (plans && plans.length > 0) {
                for (const p of plans) {
                    allPlans.push({ 
                        id: p.id, 
                        label: p.label, 
                        payment: p.payment, 
                        inventory: p.inventory, 
                        benefits: safeParse(p.benefits), 
                        rewards: safeParse(p.rewards) 
                    });
                }
            }
        } catch (e) {
            console.error("[vanguard_esc] Error fetching plans via SQL:", e.message);
        }

        return {
            list,
            allPlans,
            stats: { total: onlineCount + offlineCount, online: onlineCount, offline: offlineCount }
        };
    });

    lib.callback.register('mri_esc:vip:admin:grant', async (source, data) => {
        const srcStr = source.toString();
        const nowTs = Math.floor(Date.now() / 1000);
        if (adminCooldowns[srcStr] && (nowTs - adminCooldowns[srcStr]) < 3) return { success: false, error: "Aguarde 3 segundos" };
        adminCooldowns[srcStr] = nowTs;

        if (!bridge.IsAdminPlayer(source)) return { success: false, error: "Sem permissão" };
        if (!data || !data.citizenId || !data.tier) return { success: false, error: "Dados inválidos" };

        const success = bridge.GrantVip(data.citizenId.toUpperCase(), data.tier, data.durationDays, "Admin");
        return { success };
    });

    lib.callback.register('mri_esc:vip:admin:revoke', async (source, data) => {
        if (!bridge.IsAdminPlayer(source)) return { success: false, error: "Sem permissão" };
        if (!data || !data.citizenId) return { success: false, error: "ID obrigatório" };
        const success = bridge.RevokeVip(data.citizenId.toUpperCase(), 'admin');
        return { success };
    });

    lib.callback.register('mri_esc:vip:admin:extend', async (source, data) => {
        if (!bridge.IsAdminPlayer(source)) return { success: false, error: "Sem permissão" };
        const success = bridge.ExtendVip(data.citizenId.toUpperCase(), data.tier, data.days, "Admin");
        return { success };
    });

    lib.callback.register('mri_esc:vip:admin:search', async (source, data) => {
        if (!bridge.IsAdminPlayer(source)) return [];
        if (!data || !data.query || data.query.length < 2) return [];

        const search = data.query.toLowerCase();
        const players = exports.qbx_core.GetQBPlayers();
        let results = Object.values(players)
            .map(p => ({
                citizenid: p.PlayerData.citizenid,
                name: `${p.PlayerData.charinfo.firstname} ${p.PlayerData.charinfo.lastname}`,
                online: true,
                vip: p.PlayerData.metadata['vip'] || 'nenhum'
            }))
            .filter(p => p.name.toLowerCase().includes(search) || p.citizenid.toLowerCase().includes(search));

        try {
            const rows = await exports.oxmysql.query(`
                SELECT p.citizenid,
                    CONCAT(JSON_UNQUOTE(JSON_EXTRACT(p.charinfo,'$.firstname')),' ',
                            JSON_UNQUOTE(JSON_EXTRACT(p.charinfo,'$.lastname'))) AS name,
                    JSON_UNQUOTE(JSON_EXTRACT(p.metadata,'$.vip')) AS vip
                FROM players p
                WHERE p.citizenid LIKE ? OR JSON_UNQUOTE(JSON_EXTRACT(p.charinfo,'$.firstname')) LIKE ?
            `, [`%${search}%`, `%${search}%`]);

            if (rows) {
                const seen = new Set(results.map(r => r.citizenid));
                for (const row of rows) {
                    if (!seen.has(row.citizenid)) {
                        results.push({ citizenid: row.citizenid, name: row.name, online: false, vip: row.vip || 'nenhum' });
                    }
                }
            }
        } catch (e) {}
        return results;
    });

    console.log("[vanguard_esc] ADMIN CONTROLLER JS RESTORED SUCCESSFULLY");
}
